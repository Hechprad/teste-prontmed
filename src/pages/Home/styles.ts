import styled from 'styled-components'

import { Text as TextCmp } from 'components'

export const Wrapper = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
`

export const Title = styled(TextCmp)`
  padding: ${({ theme: { spacing } }) => spacing(2)}px;
`
