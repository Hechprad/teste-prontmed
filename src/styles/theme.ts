import { DefaultTheme as StyledDefaultTheme } from 'styled-components'

declare module 'styled-components' {
  export interface DefaultTheme {
    colors: { [key in keyof typeof themeColors]: string }
    fontTypes: { [key in keyof typeof fontTypes]: string }
    shadows: { [key in keyof typeof shadows]: string }
    spacing: (px: number) => number
  }
}

export const themeColors = {
  black: '#0F0F0F',
  blue: '#4989b5',
  gray1: '#E7E7E7',
  gray2: '#9f9f9f',
  green: '#55B046',
  red: '#F53C3C',
  white: '#FFFFFF',
}

export const shadows = {
  shadow1:
    '0px 0px 1px 1px rgba(53, 124, 173, 0.2), 0px 6px 14px 2px rgba(53, 124, 173, 0.06)',
}

export const fontTypes = {
  heading: ` font-family: Nunito Sans Bold; font-size: 24; line-height: 33.6px;`,
  body: `font-family: Nunito Sans Regular; font-size: 16; font-weight: 600; line-height: 22.4px;`,
  subtitle: `font-family: Nunito Sans Bold; font-size: 14; font-weight: 700; line-height: 22.4px;`,
}

const theme: StyledDefaultTheme = {
  colors: themeColors,
  fontTypes,
  shadows,
  spacing: (number) => number * 8,
}

export default theme
