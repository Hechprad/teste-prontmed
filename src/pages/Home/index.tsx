import React from 'react'
import { v4 } from 'uuid'

import useGetTodos from 'hooks/useGetTodos'
import useCheckTodo from 'hooks/useCheckTodo'
import useDeleteTodo from 'hooks/useDeleteTodo'
import { TodoInterface } from 'store/modules/todos/types'
import { Button, Menu, Text, TodoItem } from 'components'

import AddDropdown from './Dropdowns/AddDropdown'
import EditDropdown from './Dropdowns/EditDropdown'
import SearchDropdown from './Dropdowns/SearchDropdown'

import { ButtonWrapper, Content, Loading, Title, Ul } from './styles'

const Home: React.FC = () => {
  const { checkTodo } = useCheckTodo()
  const { deleteTodo } = useDeleteTodo()
  const { todos, loadTodos, isLoading, hasError } = useGetTodos()
  const [isAddOpen, setIsAddOpen] = React.useState<boolean>(false)
  const [isEditOpen, setIsEditOpen] = React.useState<boolean>(false)
  const [isSearchOpen, setIsSearchOpen] = React.useState<boolean>(false)
  const [selectedTodo, setSelectedTodo] = React.useState<TodoInterface>({
    id: 0,
    name: '',
    completed: false,
    inserted_at: '',
    updated_at: '',
  })

  // get todos list
  React.useEffect(() => {
    if (!todos.length && !hasError && !isLoading) {
      loadTodos()
    }
  }, [loadTodos, todos, hasError, isLoading])

  const renderList = React.useMemo(
    () => (
      <Ul>
        {todos
          .sort((a, b) => +a.completed - +b.completed)
          .map((todo) => (
            <TodoItem
              key={v4()}
              id={todo.id}
              title={todo.name}
              checked={todo.completed}
              handleEditClick={() => {
                setIsEditOpen(true)
                setSelectedTodo(todo)
              }}
              handleCheckClick={() => checkTodo(todo)}
              handleDeleteClick={() => deleteTodo(todo.id)}
            />
          ))}
      </Ul>
    ),
    [checkTodo, deleteTodo, todos]
  )

  const renderTryAgainButton = () => (
    <ButtonWrapper>
      <Text>Ops...</Text>
      <Button buttonName="Try again" onClick={() => loadTodos()} />
    </ButtonWrapper>
  )

  const renderLoading = () => <Loading />

  const getContent = () => {
    if (isLoading) {
      return renderLoading()
    }

    if (!isLoading && hasError) {
      return renderTryAgainButton()
    }

    if (!isLoading && !hasError && todos.length) {
      return renderList
    }

    return renderLoading()
  }

  return (
    <>
      <Content>
        <Title type="heading">All Tasks</Title>
        <Menu
          handleAddOpen={() => setIsAddOpen(true)}
          handleSearchOpen={() => setIsSearchOpen(true)}
        />
        {getContent()}
      </Content>
      <AddDropdown isOpen={isAddOpen} handleClose={() => setIsAddOpen(false)} />
      <SearchDropdown
        isOpen={isSearchOpen}
        handleClose={() => setIsSearchOpen(false)}
      />
      <EditDropdown
        isOpen={isEditOpen}
        selectedTodo={selectedTodo}
        handleClose={() => setIsEditOpen(false)}
      />
    </>
  )
}

export default Home
