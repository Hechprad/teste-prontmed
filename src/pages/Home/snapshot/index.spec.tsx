import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import createStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import { ThemeProvider } from 'styled-components'
import { BrowserRouter } from 'react-router-dom'

import theme from 'styles/theme'

import Home from '../index'

const INITIAL_STATE = {
  todosList: {
    todos: [
      {
        completed: true,
        id: 1,
        inserted_at: '2021-03-09T22:00:55.111320',
        name: 'new todo 1',
        updated_at: '2021-03-09T22:34:55.111320',
      },
      {
        completed: false,
        id: 2,
        inserted_at: '2021-03-09T22:34:55.111320',
        name: 'new todo 2',
        updated_at: '2021-03-09T22:34:55.111320',
      },
    ],
    isLoading: false,
    hasError: false,
  },
}
const mockStore = createStore([])
const store = mockStore(INITIAL_STATE)

beforeAll(() => {
  ReactDOM.createPortal = jest.fn((element) => element as React.ReactPortal)
})

test('Home should renders correctly', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <BrowserRouter>
            <Home />
          </BrowserRouter>
        </ThemeProvider>
      </Provider>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})
