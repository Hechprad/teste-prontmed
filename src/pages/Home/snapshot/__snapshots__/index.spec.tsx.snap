// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should renders correctly 1`] = `
Array [
  <div
    className="sc-giIncl kepWOA"
  >
    <h1
      className="sc-bdfBwQ dbsWih sc-ezrdKe loaAtj"
      color="black"
      type="heading"
    >
      All Tasks
    </h1>
    <div
      className="sc-fubCfw dWVklU"
    >
      <div
        className="sc-pFZIQ kCMVVB"
        onClick={[Function]}
      >
        <svg
          className="sc-kEjbxe eCxaPN"
          fill="none"
          height="1em"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="11"
            cy="11"
            r="8"
          />
          <line
            x1="21"
            x2="16.65"
            y1="21"
            y2="16.65"
          />
        </svg>
        <p
          className="sc-bdfBwQ fvAVWZ"
          color="blue"
          type="subtitle"
        >
          Search
        </p>
      </div>
      <div
        className="sc-pFZIQ kCMVVB"
        onClick={[Function]}
      >
        <svg
          className="sc-jrAGrp bOZxyS"
          fill="none"
          height="1em"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <line
            x1="12"
            x2="12"
            y1="8"
            y2="16"
          />
          <line
            x1="8"
            x2="16"
            y1="12"
            y2="12"
          />
        </svg>
        <p
          className="sc-bdfBwQ fvAVWZ"
          color="blue"
          type="subtitle"
        >
          Add todo
        </p>
      </div>
    </div>
    <ul
      className="sc-bYEvPH diUALu"
    >
      <li
        className="sc-crrsfI nHtIm"
      >
        <div
          className="sc-dQppl dJoByg"
        >
          <svg
            checked={false}
            className="sc-kstrdz fkbxLL"
            fill="none"
            height="1em"
            onClick={[Function]}
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M22 11.08V12a10 10 0 1 1-5.93-9.14"
            />
            <polyline
              points="22 4 12 14.01 9 11.01"
            />
          </svg>
          <p
            className="sc-bdfBwQ deLoMA"
            color="black"
            type="body"
          >
            ID: 
            2
             - 
            new todo 2
          </p>
        </div>
        <div
          className="sc-bqyKva joiINn"
        >
          <svg
            className="sc-fodVxV eHnlQc"
            fill="none"
            height="1em"
            onClick={[Function]}
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 20h9"
            />
            <path
              d="M16.5 3.5a2.121 2.121 0 0 1 3 3L7 19l-4 1 1-4L16.5 3.5z"
            />
          </svg>
          <svg
            className="sc-hBEYos licVqv"
            fill="none"
            height="1em"
            onClick={[Function]}
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="3 6 5 6 21 6"
            />
            <path
              d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
            />
            <line
              x1="10"
              x2="10"
              y1="11"
              y2="17"
            />
            <line
              x1="14"
              x2="14"
              y1="11"
              y2="17"
            />
          </svg>
        </div>
      </li>
      <li
        className="sc-crrsfI nHtIm"
      >
        <div
          className="sc-dQppl dJoByg"
        >
          <svg
            checked={true}
            className="sc-kstrdz clxhPx"
            fill="none"
            height="1em"
            onClick={[Function]}
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M22 11.08V12a10 10 0 1 1-5.93-9.14"
            />
            <polyline
              points="22 4 12 14.01 9 11.01"
            />
          </svg>
          <p
            className="sc-bdfBwQ deLoMA"
            color="black"
            type="body"
          >
            ID: 
            1
             - 
            new todo 1
          </p>
        </div>
        <div
          className="sc-bqyKva joiINn"
        >
          <svg
            className="sc-fodVxV eHnlQc"
            fill="none"
            height="1em"
            onClick={[Function]}
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 20h9"
            />
            <path
              d="M16.5 3.5a2.121 2.121 0 0 1 3 3L7 19l-4 1 1-4L16.5 3.5z"
            />
          </svg>
          <svg
            className="sc-hBEYos licVqv"
            fill="none"
            height="1em"
            onClick={[Function]}
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="3 6 5 6 21 6"
            />
            <path
              d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
            />
            <line
              x1="10"
              x2="10"
              y1="11"
              y2="17"
            />
            <line
              x1="14"
              x2="14"
              y1="11"
              y2="17"
            />
          </svg>
        </div>
      </li>
    </ul>
  </div>,
  <div
    className="sc-dlfnbm cEmwKr"
  >
    <div
      className="sc-hKgILt hNSNgx"
      onClick={[Function]}
    />
    <div
      className="sc-eCssSg fdbUnc"
      size="sm"
    >
      <div
        className="sc-gKsewC dElHde"
      >
        <p
          className="sc-bdfBwQ deLoMA sc-bkzZxe jGCbDc"
          color="black"
          type="body"
        >
          Add new todo
        </p>
      </div>
      <div
        className="sc-jSgupP kDIAus"
      >
        <div
          className="sc-fFubgz bqIaKo"
        >
          <input
            className="sc-iqHYGH egqCRi"
            onChange={[Function]}
            placeholder="New todo name"
            value=""
          />
          <button
            className="sc-gsTCUz ehKRyO sc-dIUggk idQFbO"
            disabled={true}
            onClick={[Function]}
          >
            <svg
              className="sc-hHftDr gQJrLR"
              fill="none"
              height="1em"
              stroke="currentColor"
              strokeLinecap="round"
              strokeLinejoin="round"
              strokeWidth="2"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="10"
              />
              <line
                x1="12"
                x2="12"
                y1="8"
                y2="16"
              />
              <line
                x1="8"
                x2="16"
                y1="12"
                y2="12"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>,
  <div
    className="sc-dlfnbm cEmwKr"
  >
    <div
      className="sc-hKgILt hNSNgx"
      onClick={[Function]}
    />
    <div
      className="sc-eCssSg fdbUnc"
      size="sm"
    >
      <div
        className="sc-gKsewC dElHde"
      >
        <p
          className="sc-bdfBwQ deLoMA sc-cxFLnm fyvzqG"
          color="black"
          type="body"
        >
          Search todo by ID
        </p>
      </div>
      <div
        className="sc-jSgupP kDIAus"
      >
        <div
          className="sc-iwyYcG ehMonG"
        >
          <input
            className="sc-iqHYGH egqCRi sc-lmoMRL dislZm"
            onChange={[Function]}
            placeholder="TODO ID"
            value=""
          />
          <button
            className="sc-gsTCUz ehKRyO sc-iJuUWI bKPCkQ"
            disabled={true}
            onClick={[Function]}
          >
            Search
          </button>
        </div>
        <div
          className="sc-iwyYcG ehMonG"
        />
      </div>
    </div>
  </div>,
  <div
    className="sc-dlfnbm cEmwKr"
  >
    <div
      className="sc-hKgILt hNSNgx"
      onClick={[Function]}
    />
    <div
      className="sc-eCssSg fdbUnc"
      size="sm"
    >
      <div
        className="sc-gKsewC dElHde"
      >
        <p
          className="sc-bdfBwQ deLoMA sc-kfzAmx glnuDD"
          color="black"
          type="body"
        >
          Update todo name
        </p>
      </div>
      <div
        className="sc-jSgupP kDIAus"
      >
        <div
          className="sc-dmlrTW fsEKZe"
        >
          <p
            className="sc-bdfBwQ deLoMA"
            color="black"
            type="body"
          >
            Todo ID: 
            0
          </p>
          <input
            className="sc-iqHYGH egqCRi"
            onChange={[Function]}
            placeholder="New name"
            value=""
          />
          <button
            className="sc-gsTCUz ehKRyO sc-bBXqnf fOVJdl"
            disabled={true}
            onClick={[Function]}
          >
            Update
          </button>
        </div>
      </div>
    </div>
  </div>,
]
`;
